File: packages/zod/src/research.ts
import { z } from "zod";

export const ZResearchStatus = z.enum([
  "queued",
  "running",
  "completed",
  "failed",
  "partial",
]);

export const ZResearchRequest = z.object({
  query: z.string().min(10).max(500),
  session_id: z.string().uuid().optional(),
});

export const ZReportSection = z.object({
  heading: z.string(),
  content: z.string(),
  citations: z.array(z.string()),
});

export const ZReportSource = z.object({
  title: z.string(),
  url: z.string().url(),
  source_type: z.enum(["web", "pdf", "internal_doc"]),
});

export const ZReportConfidence = z.object({
  overall: z.number().min(0).max(1),
  data_quality: z.number().min(0).max(1),
  source_reliability: z.number().min(0).max(1),
});

export const ZResearchEvent = z.object({
  type: z.string(),
  status: ZResearchStatus,
  message: z.string(),
  agent: z.string().optional(),
  timestamp: z.string().datetime(),
  data: z.record(z.string(), z.any()).optional(),
});

export const ZResearchReport = z.object({
  title: z.string(),
  summary: z.string(),
  sections: z.array(ZReportSection),
  sources: z.array(ZReportSource),
  confidence: ZReportConfidence,
});

export const ZResearchCreatedResponse = z.object({
  session_id: z.string().uuid(),
  status: ZResearchStatus,
  stream_url: z.string(),
});

export const ZResearchSessionResponse = z.object({
  session_id: z.string().uuid(),
  query: z.string(),
  status: ZResearchStatus,
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  started_at: z.string().datetime().optional(),
  completed_at: z.string().datetime().optional(),
  report: ZResearchReport.optional(),
  events: z.array(ZResearchEvent).optional(),
  error: z.string().optional(),
});

File: app/backend/internal/model/research.go
package model

import "time"

type ResearchStatus string

const (
	ResearchStatusQueued    ResearchStatus = "queued"
	ResearchStatusRunning   ResearchStatus = "running"
	ResearchStatusCompleted ResearchStatus = "completed"
	ResearchStatusFailed    ResearchStatus = "failed"
	ResearchStatusPartial   ResearchStatus = "partial"
)

type ResearchRequest struct {
	Query     string `json:"query" validate:"required,min=10,max=500"`
	SessionID string `json:"session_id,omitempty" validate:"omitempty,uuid"`
}

type ReportConfidence struct {
	Overall           float64 `json:"overall"`
	DataQuality       float64 `json:"data_quality"`
	SourceReliability float64 `json:"source_reliability"`
}

type ReportSection struct {
	Heading   string   `json:"heading"`
	Content   string   `json:"content"`
	Citations []string `json:"citations"`
}

type ReportSource struct {
	Title      string `json:"title"`
	URL        string `json:"url"`
	SourceType string `json:"source_type"`
}

type ResearchReport struct {
	Title      string           `json:"title"`
	Summary    string           `json:"summary"`
	Sections   []ReportSection  `json:"sections"`
	Sources    []ReportSource   `json:"sources"`
	Confidence ReportConfidence `json:"confidence"`
}

type ResearchEvent struct {
	Type      string         `json:"type"`
	Status    ResearchStatus `json:"status"`
	Message   string         `json:"message"`
	Agent     string         `json:"agent,omitempty"`
	Timestamp time.Time      `json:"timestamp"`
	Data      map[string]any `json:"data,omitempty"`
}

type ResearchSession struct {
	SessionID   string          `json:"session_id"`
	Query       string          `json:"query"`
	Status      ResearchStatus  `json:"status"`
	CreatedAt   time.Time       `json:"created_at"`
	UpdatedAt   time.Time       `json:"updated_at"`
	StartedAt   *time.Time      `json:"started_at,omitempty"`
	CompletedAt *time.Time      `json:"completed_at,omitempty"`
	Report      *ResearchReport `json:"report,omitempty"`
	Events      []ResearchEvent `json:"events,omitempty"`
	Error       string          `json:"error,omitempty"`
}

type ResearchCreatedResponse struct {
	SessionID string         `json:"session_id"`
	Status    ResearchStatus `json:"status"`
	StreamURL string         `json:"stream_url"`
}
